src/i_midimusic.c

```diff
@@ -1383,287 +1148,261 @@ static boolean IsRPGLoop(void)
 
-static DWORD WINAPI PlayerProc(LPVOID lpParam)
+    int64_t us = ((int64_t)delta_time * us_per_beat) / ticks_per_beat;
+    *wait_time = I_GetTimeUS() + us;
+    return STATE_WAITING;
+}
+
+static int PlayerThread(void *unused)
 {
-    boolean keep_going = true;
+    SDL_SetThreadPriority(SDL_THREAD_PRIORITY_TIME_CRITICAL);
+
+    midi_event_t *event = NULL;
+    midi_track_t *track = NULL;
+    boolean sleep = false;
+    int64_t wait_time = 0;
 
-    while (keep_going)
+    while (SDL_AtomicGet(&player_thread_running))
     {
-        if (WaitForSingleObject(hBufferReturnEvent, INFINITE) != WAIT_OBJECT_0)
+        if (sleep)
         {
-            continue;
+            sleep = false;
+            I_Sleep(1);
         }
 
         // The MIDI thread must have exclusive access to shared resources until
         // the end of the current loop iteration or when the thread exits.
-        EnterCriticalSection(&CriticalSection);
 
-        switch (win_midi_state)
+        SDL_LockMutex(music_lock);
+
+        switch (midi_state)
         {
             case STATE_STARTUP:
                 ResetDevice();
-                StreamOut();
                 song.rpg_loop = IsRPGLoop();
-                win_midi_state = STATE_PLAYING;
-                break;
-
-            case STATE_SHUTDOWN:
-                // Send notes/sound off prior to reset to prevent volume spikes.
-                SendNotesSoundOff();
-                ResetDevice();
-                StreamOut();
-                win_midi_state = STATE_EXIT;
-                break;
-
-            case STATE_EXIT:
-                keep_going = false;
+                midi_state = STATE_PLAYING;
                 break;
 
             case STATE_PLAYING:
-                if (update_volume)
-                {
-                    UpdateVolume();
-                    StreamOut();
-                    update_volume = false;
-                    break;
-                }
-                FillBuffer();
+                midi_state = NextEvent(&event, &track, &wait_time);
                 break;
 
-            case STATE_STOPPING:
+            case STATE_PAUSING:
                 // Send notes/sound off to prevent hanging notes.
                 SendNotesSoundOff();
-                StreamOut();
-                win_midi_state = STATE_STOPPED;
+                midi_state = STATE_PAUSED;
                 break;
 
-            case STATE_STOPPED:
-                SetEvent(hStoppedEvent);
+            case STATE_WAITING:
+                {
+                    int64_t remaining_time = wait_time - I_GetTimeUS();
+                    if (remaining_time <= 0)
+                    {
+                        ProcessEvent(event, track);
+                        midi_state = STATE_PLAYING;
+                    }
+                    else if (remaining_time > 1000)
+                    {
+                        sleep = true;
+                    }
+                }
                 break;
```

Comment:
So if there is 1-1000 us remaining, the loop runs at full speed for that duration? Will that thrash the CPU?

Reply:
Fixed in https://github.com/fabiangreffrath/woof/pull/1635/commits/d5dcf60e245a3ccff7d3062bc1d583808c19ff09